File: src/server/scripts/demo01_server.py
#! /usr/bin/env python

import rospy
from server.srv import AddInts,AddIntsRequest,AddIntsResponse


def doNums(request):
    """
    处理客户端的AddIntsRequest消息。
    参数：封装了请求数据
    返回值：响应数据
    """
    # 解析提交的数
    num1 = request.num1
    num2 = request.num2

    # 计算并返回结果
    sum = num1 + num2

    # 将结果封装进响应
    response = AddIntsResponse()
    response.sum = sum

    rospy.loginfo(f"服务器解析的数据: num1 {num1}, num2 {num2}, sum {sum}")
    return response

    

def main():
    """
    服务端：解析客户端请求，产生响应。
    """
    rospy.init_node("server")

    # 创建服务端对象
    server = rospy.Service("addInts",AddInts,doNums)
    rospy.loginfo("服务器启动")
    rospy.spin()
if __name__ == "__main__":
    main()

File: src/server/src/demo01_server.cpp
#include "ros/ros.h"
#include "server/AddInts.h"


bool doNums(server::AddInts::Request& request,
            server::AddInts::Response &response){
    /*
    request: 请求对象
    response：响应对象
    */

   //1.处理请求
   int num1 = request.num1;
   int num2 = request.num2;
   ROS_INFO("num1: %d, num2: %d", num1,num2);

   //2.组织响应
    int sum = num1 + num2;
    response.sum = sum;
    ROS_INFO("sum: %d",sum);

    return true;
}

int main(int argc, char *argv[]){

    setlocale(LC_ALL, "");
    
    ros::init(argc, argv,"server");
    ROS_INFO("服务器启动");
    ros::NodeHandle nh;

    //和topic不同，这里需要创建服务对象，并且处理请求并产生响应。
    ros::ServiceServer server = nh.advertiseService("addInts",doNums);
    
    ros::spin();

    return 0;
}